File: app/src/main/java/com/suzei/racoon/ui/base/Contract.java
package com.suzei.racoon.ui.base;

import androidx.recyclerview.widget.RecyclerView;

public interface Contract {

    interface ProgressView {

        void showProgress();

        void hideProgress();

    }

    interface DetailsView<M> extends ProgressView {

        void onLoadSuccess(M data);

        void onLoadFailed(String message);

    }

    interface AdapterView<A extends RecyclerView.Adapter> extends ProgressView {

        void setAdapter(A adapter);

        void loadFailed();

    }

    interface Listener<M> {

        void onLoadSuccess(M data);

        void onLoadFailed(String message);

    }

    interface AdapterListener<A extends RecyclerView.Adapter> {

        void onLoadSuccess(A data);

        void onLoadFailed();

    }

}

File: app/src/main/java/com/suzei/racoon/ui/base/Contract.kt
package com.suzei.racoon.ui.base

import androidx.recyclerview.widget.RecyclerView

interface Contract {
    interface ProgressView {
        fun showProgress()
        fun hideProgress()
    }

    interface DetailsView<M> : ProgressView {
        fun onLoadSuccess(data: M)
        fun onLoadFailed(message: String)
    }

    interface AdapterView<A : RecyclerView.Adapter<*>> : ProgressView {
        fun setAdapter(adapter: A)
        fun loadFailed()
    }

    interface Listener<M> {
        fun onLoadSuccess(data: M)
        fun onLoadFailed(message: String)
    }

    interface AdapterListener<A : RecyclerView.Adapter<*>> {
        fun onLoadSuccess(data: A)
        fun onLoadFailed()
    }
}